jax/random.py

```diff
@@ -1224,8 +1223,36 @@ def logistic(key, shape=(), dtype=onp.float64):
 
 @partial(jit, static_argnums=(1, 2))
 def _logistic(key, shape, dtype):
+  # Mathematically, we should draw a uniform variable x in [0, 1], and
+  # compute the logistic random variable using log(x / (1 - x)).
+  # An issue arises when implementing this directly via uniform(), because
+  #  1) uniform() generates values in the semi-closed interval [0, 1)
+  #  2) when x=0, the output is -inf
+  #  3) uniform() only randomizes the mantissa, meaning that for float32 input
+  #     the cardinality is only about 2^23 which means these infinities are quite
+  #     common (about 1 in 8 million)
+  # We can address this by generating integers I in (0, N), then transform
+  # using log(I / (N - I)). Note that the endpoints here are exclusive, so
+  # infinities are avoided at either end.
+  # But what should we choose for N? It's most efficient to generate integers
+  # in a range [0, 2^M) for integer M. We can add 1 to get (0, 2^M + 1).
+  # We cannot use M = nbits because 2^nbits overflows, so we use M = nbits - 1
+  # to get the highest possible cardinality in the most efficient manner.
+  # (Note: we avoid randint because it generates 2x more bits than necessary).
   _check_shape("logistic", shape)
-  return logit(uniform(key, shape, dtype))
+  nbits = np.finfo(dtype).bits
```

Comment:
Should this function accept complex `dtype`s? I think it does not at the moment, but I believe it would after your change.

Reply:
Done.